caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Managed TGIS subprocess with automatic failure recovery
+
+Creates a wrapped gRPC client that knows that the gRPC server is managed as a
+subprocess. When the client observes errors, it can trigger a restart of the
+subprocess.
+"""
+
+# Standard
+from enum import Enum
+from threading import Lock
+from typing import Optional
+import os
+import shlex
+import subprocess
+import threading
+import time
+
+# Third Party
+import grpc
+import requests
+
+# First Party
+from caikit.core.toolkit.errors import error_handler
+import alog
+
+# Local
+from .protobufs import generation_pb2_grpc
+
+log = alog.use_channel("TGISPROC")
+error = error_handler.get(log)
+
+
+class _TGISState(Enum):
+    STOPPED = 0  # subprocess does not exist
+    BOOTING = 1  # attempting to load the model
+    READY = 2  # ready for inference requests
+
+
+# pylint: disable=too-many-instance-attributes
+class ManagedTGISSubprocess:
```

Comment:
NIT: It would be nice to have a docstring. If this should be the same as the module-level docstring, you can cheat by doing the following:

```py
class ManagedTGISSubproces:
    __doc__ = __doc__
    ...
```